File: scripts/compile-decomp.py
#!/usr/bin/env python3
"""
Compile Make Me a Hanzi dictionary.txt into a binary format for Zig @embedFile().

Binary format:
  Header (8 bytes):
    magic: "DCMP" (4 bytes)
    count: u32 LE (number of entries)

  Offset index (count * 4 bytes):
    offset[i]: u32 LE — byte offset of entry i in the entry table

  Entry table (variable-length entries, sorted by character UTF-8):
    Each entry:
      char_len: u8
      radical_len: u8
      decomposition_len: u8
      pinyin_len: u8
      definition_len: u16 LE
      etymology_type_len: u8
      etymology_hint_len: u8
      char: [char_len]u8
      radical: [radical_len]u8
      decomposition: [decomposition_len]u8
      pinyin: [pinyin_len]u8
      definition: [definition_len]u8
      etymology_type: [etymology_type_len]u8
      etymology_hint: [etymology_hint_len]u8

Usage:
  python3 scripts/compile-decomp.py /tmp/dictionary.txt libtaijobi/src/decomp.bin
"""

import json
import struct
import sys


def parse_dictionary(path: str) -> list[dict]:
    """Parse Make Me a Hanzi dictionary.txt (NDJSON)."""
    entries = []
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue

            character = obj.get('character', '')
            if not character:
                continue

            radical = obj.get('radical', '')
            decomposition = obj.get('decomposition', '')
            pinyin_list = obj.get('pinyin', [])
            pinyin = ', '.join(pinyin_list) if isinstance(pinyin_list, list) else str(pinyin_list or '')
            definition = obj.get('definition', '') or ''

            # Etymology
            etymology = obj.get('etymology', {}) or {}
            ety_type = etymology.get('type', '') or ''
            ety_hint = etymology.get('hint', '') or ''

            entries.append({
                'character': character,
                'radical': radical,
                'decomposition': decomposition,
                'pinyin': pinyin,
                'definition': definition,
                'etymology_type': ety_type,
                'etymology_hint': ety_hint,
            })

    return entries


def compile_binary(entries: list[dict], out_path: str):
    """Compile entries into binary format."""
    # Sort by character (UTF-8 byte order)
    entries.sort(key=lambda e: e['character'].encode('utf-8'))

    entry_data = bytearray()
    offsets = []

    for e in entries:
        c = e['character'].encode('utf-8')
        r = e['radical'].encode('utf-8')
        d = e['decomposition'].encode('utf-8')
        p = e['pinyin'].encode('utf-8')
        df = e['definition'].encode('utf-8')
        et = e['etymology_type'].encode('utf-8')
        eh = e['etymology_hint'].encode('utf-8')

        # Truncate to fit length fields
        if len(c) > 255 or len(r) > 255 or len(d) > 255 or len(p) > 255:
            continue
        if len(df) > 65535:
            df = df[:65535]
        if len(et) > 255 or len(eh) > 255:
            et = et[:255]
            eh = eh[:255]

        offsets.append(len(entry_data))

        # Header: char_len, radical_len, decomp_len, pinyin_len, def_len(u16), ety_type_len, ety_hint_len
        entry_data += struct.pack('<BBBBHBB',
            len(c), len(r), len(d), len(p), len(df), len(et), len(eh))
        entry_data += c + r + d + p + df + et + eh

    count = len(offsets)

    with open(out_path, 'wb') as f:
        # Header
        f.write(b'DCMP')
        f.write(struct.pack('<I', count))

        # Offset index
        for off in offsets:
            f.write(struct.pack('<I', off))

        # Entry table
        f.write(entry_data)

    print(f"Compiled {count} entries -> {out_path}")
    total_size = 8 + count * 4 + len(entry_data)
    print(f"Total size: {total_size / 1024 / 1024:.1f} MB")


def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <dictionary.txt> <output.bin>")
        sys.exit(1)

    entries = parse_dictionary(sys.argv[1])
    print(f"Parsed {len(entries)} entries")
    compile_binary(entries, sys.argv[2])


if __name__ == '__main__':
    main()

File: scripts/compile-decomp.ts
#!/usr/bin/env bun
/**
 * Compile Make Me a Hanzi dictionary.txt into a binary format for Zig @embedFile().
 *
 * Binary format:
 *   Header (8 bytes):
 *     magic: "DCMP" (4 bytes)
 *     count: u32 LE (number of entries)
 *
 *   Offset index (count * 4 bytes):
 *     offset[i]: u32 LE — byte offset of entry i in the entry table
 *
 *   Entry table (variable-length entries, sorted by character UTF-8):
 *     Each entry:
 *       char_len: u8
 *       radical_len: u8
 *       decomposition_len: u8
 *       pinyin_len: u8
 *       definition_len: u16 LE
 *       etymology_type_len: u8
 *       etymology_hint_len: u8
 *       char: [char_len]u8
 *       radical: [radical_len]u8
 *       decomposition: [decomposition_len]u8
 *       pinyin: [pinyin_len]u8
 *       definition: [definition_len]u8
 *       etymology_type: [etymology_type_len]u8
 *       etymology_hint: [etymology_hint_len]u8
 *
 * Usage:
 *   bun scripts/compile-decomp.js <dictionary.txt> <output.bin>
 */

import { readFileSync, writeFileSync } from "node:fs";

const encoder = new TextEncoder();

function parseDictionary(path: string) {
  const text = readFileSync(path, "utf-8");
  const entries = [];

  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    let obj;
    try {
      obj = JSON.parse(trimmed);
    } catch {
      continue;
    }

    const character = obj.character || "";
    if (!character) continue;

    const radical = obj.radical || "";
    const decomposition = obj.decomposition || "";
    const pinyinList = obj.pinyin;
    const pinyin = Array.isArray(pinyinList)
      ? pinyinList.join(", ")
      : String(pinyinList || "");
    const definition = obj.definition || "";

    const etymology = obj.etymology || {};
    const etymologyType = etymology.type || "";
    const etymologyHint = etymology.hint || "";

    entries.push({
      character,
      radical,
      decomposition,
      pinyin,
      definition,
      etymologyType,
      etymologyHint,
    });
  }

  return entries;
}

function compileBinary(entries, outPath) {
  // Sort by character UTF-8 byte order
  entries.sort((a, b) => {
    const ab = encoder.encode(a.character);
    const bb = encoder.encode(b.character);
    const len = Math.min(ab.length, bb.length);
    for (let i = 0; i < len; i++) {
      if (ab[i] !== bb[i]) return ab[i] - bb[i];
    }
    return ab.length - bb.length;
  });

  const entryBuffers = [];
  const offsets = [];
  let entryDataLen = 0;

  for (const e of entries) {
    const c = encoder.encode(e.character);
    const r = encoder.encode(e.radical);
    const d = encoder.encode(e.decomposition);
    const p = encoder.encode(e.pinyin);
    let df = encoder.encode(e.definition);
    let et = encoder.encode(e.etymologyType);
    let eh = encoder.encode(e.etymologyHint);

    // Truncate to fit length fields
    if (c.length > 255 || r.length > 255 || d.length > 255 || p.length > 255) {
      continue;
    }
    if (df.length > 65535) {
      df = df.slice(0, 65535);
    }
    if (et.length > 255) {
      et = et.slice(0, 255);
    }
    if (eh.length > 255) {
      eh = eh.slice(0, 255);
    }

    offsets.push(entryDataLen);

    // Header: char_len(u8), radical_len(u8), decomp_len(u8), pinyin_len(u8),
    //         def_len(u16 LE), ety_type_len(u8), ety_hint_len(u8)
    const header = new Uint8Array(8);
    header[0] = c.length;
    header[1] = r.length;
    header[2] = d.length;
    header[3] = p.length;
    header[4] = df.length & 0xff;
    header[5] = (df.length >> 8) & 0xff;
    header[6] = et.length;
    header[7] = eh.length;

    const entryLen = 8 + c.length + r.length + d.length + p.length + df.length + et.length + eh.length;
    const buf = new Uint8Array(entryLen);
    let pos = 0;
    buf.set(header, pos); pos += 8;
    buf.set(c, pos); pos += c.length;
    buf.set(r, pos); pos += r.length;
    buf.set(d, pos); pos += d.length;
    buf.set(p, pos); pos += p.length;
    buf.set(df, pos); pos += df.length;
    buf.set(et, pos); pos += et.length;
    buf.set(eh, pos);

    entryBuffers.push(buf);
    entryDataLen += entryLen;
  }

  const count = offsets.length;
  const totalSize = 8 + count * 4 + entryDataLen;
  const output = new Uint8Array(totalSize);
  const view = new DataView(output.buffer);

  // Header
  output[0] = 0x44; // D
  output[1] = 0x43; // C
  output[2] = 0x4d; // M
  output[3] = 0x50; // P
  view.setUint32(4, count, true);

  // Offset index
  for (let i = 0; i < count; i++) {
    view.setUint32(8 + i * 4, offsets[i], true);
  }

  // Entry table
  let pos = 8 + count * 4;
  for (const buf of entryBuffers) {
    output.set(buf, pos);
    pos += buf.length;
  }

  writeFileSync(outPath, output);

  console.log(`Compiled ${count} entries -> ${outPath}`);
  console.log(`Total size: ${(totalSize / 1024 / 1024).toFixed(1)} MB`);
}

function main() {
  const args = process.argv.slice(2);
  if (args.length < 2) {
    console.log(`Usage: bun scripts/compile-decomp.ts <dictionary.txt> <output.bin>`);
    process.exit(1);
  }

  const entries = parseDictionary(args[0]);
  console.log(`Parsed ${entries.length} entries`);
  compileBinary(entries, args[1]);
}

main();
